SwiftCompilerSources/Sources/SIL/Function.swift

```diff
@@ -318,6 +328,12 @@ extension Function {
     }
   }
 
+  public var hasIllegalEffectForSpecialization: Bool {
```

Comment:
This is not a good API, because it does not say for which kind of specialization. For example, those effect kinds are fine for generic specialization. I suggest make this private in the AutodiffClosureSpecization.
Also, usually we should return true for the "good thing". How about `effectsAllowSpecialization`?